Teach naming conventions check to look for tests in src/main

nik9000 opened the pull request:
Teaches the naming conventions check to fail the build when it finds
tests in `src/main/java`. Tests in `src/main/java` aren't run and that
is bad because tests that don't run provide a false sense of
security.

Edit: As it stands now this actually creates a new similar task to the
`namingConventions` task, `namingConventionsMain` that checks
the main classes of the `buildSrc` and `test/framework` projects,
looking for tests. This checking isn't done by the same task.

nik9000 committed:
Teach check to look for tests in src/main

Teaches tha naming convetions check to fail the build when it finds
tests in src/main/java. Tests in src/main/java aren't run and that
is bad because tests that don't run provide a false sense of
security.

jasontedor commented:
It looks like `LongGCDisruptionTest` is both incorrectly named (missing 's') and in the wrong place.

nik9000 commented:
> It looks like LongGCDisruptionTest is both incorrectly named (missing 's') and in the wrong place.

Want me to fix it in this PR or a followup?

jasontedor commented:
> Want me to fix it in this PR or a followup?

This PR, and the additional check that you proposed via another channel.

nik9000 committed:
Check for poorly named classes in src/main

nik9000 commented:
> This PR, and the additional check that you proposed via another channel.

I pushed another commit that should do that.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy:
```diff
@@ -78,9 +68,19 @@ public class NamingConventionsTask extends LoggedExec {
+        if (project.sourceSets.names.contains('main')) {
+            /* If there are main classes we'll have a look at them just to make
+             * sure there aren't file named like test classes in there too. */
+            inputs.files(project.sourceSets.main.runtimeClasspath)
```
I don't think this should be necessary? The test classpath should already include the main classpath.

nik9000 replied:
Makes sense. I'll drop it.

rjernst reviewed (commented):
This seems to complicate the naming conventions check, for an extreme edgecase. It is only possible that such an error would even compile if in the test framework, as LongGCDisruptionTest is.  I think it would be far simpler to have a separate task in test framework to look for such cases? I don't even know if it needs to be more than just looking at the classes that exist in test framework, and seeing if any end in Test, Tests, or IT.

nik9000 committed:
Drop dependency on main classpath

It is included in the test classpath.

nik9000 commented:
I think it is worth having the check. I don't think code review is a great mechanism for catching stuff like that. I didn't want to do anything just in test:framework because I though it'd be brittle to limit the check. And we already had the naming conventions check doing very similar things. I don't feel like looking at names is really good enough. I started out just looking at names and it missed the LongGCDisruptionTest.

All that being said, it is your build so if you want me to do something just for test:framework I'll change it. If you want me to drop the test changes and just move the classes I can do that too. Or some other thing I haven't thought of.

rjernst commented:
> I don't think code review is a great mechanism for catching stuff like that

I don't understand this comment. Where did I suggest that? All I said was it is not possible to have a subclass of ESTestCase under main, except in test framework, so there is no use having this check anywhere else.

nik9000 commented:
> I don't understand this comment. Where did I suggest that? All I said was it is not possible to have a subclass of ESTestCase under main, except in test framework, so there is no use having this check anywhere else.

Sorry. Do you think it is worth adding something so this is only run on test:framework? Like a flag on the task or detecting that the main classes have junit as a dependency? Do you want a separate task entirely?

rjernst commented:
I was suggesting an entirely separate task, in test:framework (which you already essentially have). I would not complicate the naming conventions task unless it is necessary, but I'm failing to see why it would be necessary.  I still think just checking names would be sufficient for this edge case. If you feel strongly, then I would say have a flag, and create a second task in test:framework which operates on the main classpath, and runs with this flag.

nik9000 committed:
Move main classpath testing to a separate task

The task still uses the NamingConventionCheck.java file because that
has the infrastructure for scanning the files and checking subclasses.

nik9000 committed:
Revert line break change

nik9000 commented:
@rjernst I liked the subclass checking a lot so I split the execution into two tasks, with the main class checking ones only being built on buildSrc and test:framework. You get the main class checking behavior by providing a flag to NamingConventionTask.

rjernst commented on buildSrc/build.gradle:
```diff
@@ -156,4 +156,11 @@ if (project != rootProject) {
     integTestClass = 'org.elasticsearch.test.NamingConventionsCheckBadClasses$IntegTestCase'
   }
+
+  task(namingConventionsMain, type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```
nit: can you use the style of task creation used throughout the rest of the build.gradle files, ie:
```
task namingConventionsMain(type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```

nik9000 replied:
Sure!

rjernst commented on test/framework/build.gradle:
```diff
@@ -63,3 +63,8 @@ thirdPartyAudit.excludes = [
   'org.jmock.core.Constraint',
 ]
+
+task(namingConventionsMain, type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```
ditto on nit about task creation syntax

rjernst reviewed (approved):
LGTM, one small nit.

nik9000 committed:
Merge branch 'master' into file_names

nik9000 committed:
Use normal syntax

nik9000 commented:
Thanks for reviewing @rjernst!
